feat: add the number of merge buffers used for druid emitter

TessaIO opened the pull request:
<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->

Fixes #4345 .

<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->

#### Add the current number of merge buffers used for historical, broker, and real-time.

<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->

<hr>
Add the current number of merge buffers used for historical, broker, and real-time.

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

LakshSingla commented on docs/operations/metrics.md:
```diff
@@ -61,6 +61,7 @@ Most metric values reset each emission period, as specified in `druid.monitoring
 |`query/success/count`|Number of queries successfully processed.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
 |`query/failed/count`|Number of failed queries.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
 |`query/interrupted/count`|Number of queries interrupted due to cancellation.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
+|`query/merge/buffersUsed`|number of merge buffers allocated to broker while performing groupBy merge queries.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
```
This shouldn't be "allocated to broker". As per my perception, allocated to the broker means the total number of merge buffers present there. Also, there is no such thing as "groupBy merge query". 
```suggestion
|`query/merge/buffersUsed`|Number of merge buffers used up to merge the results of group by queries.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
```

LakshSingla commented on processing/src/test/java/org/apache/druid/query/TestBufferPool.java:
```diff
@@ -147,4 +147,10 @@ public Collection<RuntimeException> getOutstandingExceptionsCreated()
   }
+
+  @Override
+  public int getUsedBufferCount()
+  {
+    return Integer.MAX_VALUE;
+  }
```
We can probably have the correct implementation for this class, even if it is a test utility.

LakshSingla reviewed (commented):
Thanks for the first contribution! 

The code style would be failing on a few files. Checkout 
a) https://github.com/apache/druid/blob/master/dev/intellij-setup.md#code-style to setup the editor to lint the code correctly
b) Running `mvn checkstyle:checkstyle` would help in confirming if the checkstyle would fail on the PR or not.

Regarding the PR:
a. I think there should be a `query/merge/totalBuffers` metric as well. 
b. Apart from this, there can be a wait time associated with each query - the total time required to acquire the merge buffers - this should be a simple change given that now we acquire all the merge buffers in a single place 
c. There should be tests for the changes made in the PR.

TessaIO commented:
Thanks for the review @LakshSingla.
I addressed all of your comments. PTAL when you have time.

> b. Apart from this, there can be a wait time associated with each query - the total time required to acquire the merge buffers - this should be a simple change given that now we acquire all the merge buffers in a single place

I didn't really get what needed to be done here. Can you elaborate more on this?

LakshSingla commented:
@TessaIO Aplogy for getting back late on this PR. The PR LGTM, can you please address the test failures. 

My original comment was related to adding more metrics around this area, such as the elapsed time between the query requesting the merge buffer, and the query getting allotted the merge buffer. That can be a part of subsequent PRs.

TessaIO commented:
@LakshSingla @cryptoe any idea on how to improve coverage to make the pipeline pass?

cryptoe commented:
@TessaIO Could you please rebase the patch. I cannot pull the logs from the older CI runs to see the failures.

TessaIO commented:
@cryptoe done!

cryptoe commented:
@TessaIO Could you please fix the CI.
